### PR TITLE
Add missing hash methods

### PR DESCRIPTION
**Pull request by straight-shoota**
A few types in stdlib define a custom equality operator `#==` but don't override `#hash` to produce matching results so that from `a == b` follows `a.hash = b.hash`.

I think the implementations are pretty self-explanatory. Only `Path` was a bit more because it uses case-insensitive comparison based on kind.

I ran stdlib specs with `raise "non-matching hash #{actual_value.class} #{@expected_value.class}" if actual_value.hash != @expected_value.hash` in `EqualExpectation` to find the offenders. It obviously only fetches ones where a positive `eq` expectation is used. In some cases the inherited hash implementation may just happen to work for the passed values, but not for others. That can't be detected like this.
Regardless, it might be worth adding this hash check to `EqualExpecation` permanently 🤔 

There are still a couple of failing cases left.

* hashes for big number types are broken (tracked in #10107) (fixed in #10121)
* `HTTP::Headers#==(other : Hash)` is a bit odd. The hash can't be copmatible with both `Hash(String, String)` *and* `Hash(String, Array(String))`. And the equality is not commutative, so I'm going to propose removing it. (fixed in #10186)
* Some private spec types also fail, but I didn't bother fixing them until we decide to make hash equality check mandatory.

### CODE REVIEW

File: src/path.cr [hunk 1]
```diff
@@ -1059,12 +1059,28 @@ struct Path
   # Path.windows("foo") <=> Path.windows("FOO") # => 0
   # ```
   def <=>(other : Path)
-    ord = @name.compare(other.@name, case_insensitive: windows?)
+    p! self.@name, self.@kind, other.@name, other.@kind, @name.compare(other.@name, case_insensitive: true), windows? || other.windows?
```
Comment (oprypin) [inline src/path.cr:1062]: p!

**Comment by bcardiff** [comment]
> Regardless, it might be worth adding this hash check to EqualExpecation permanently 

👎 Types that are not expected to be used as key or in set would be forced to implement a hash. 

I would, however, suggest to add some equality assertion that will verify hash and equality given some samples.

Something like:

```crystal
it_eq_and_hash [v1, v2], [v3, v4, v5]
```

That will validate 
* (reflexivity) v1 == v1, (simmetry) v1 == v2, v2 == 1, (transitifivity) v3 == v4, v4 == v5, v3 == v5, etc.
* that the argument represent different congruent classes v1 != v3
* v1.hash == v2.hash
* v3.hash == v4.hash == v5.hash

**Comment by straight-shoota** [comment]
@bcardiff Can we just merge this? I didn't plan to start a discussion on changing expectation behaviour in this PR.

File: src/uuid.cr [hunk 2]
```diff
@@ -176,6 +176,8 @@ struct UUID
     @bytes == other.@bytes
   end
 
+  def_hash @bytes
```
Comment (asterite) [inline src/uuid.cr:179]: Maybe we can use `def_equals_and_hash` and remove `==`?

File: src/socket/address.cr [hunk 3]
```diff
@@ -380,6 +382,8 @@ class Socket
       path == other.path
     end
 
+    def_hash path
```
Comment (asterite) [inline src/socket/address.cr:385]: Maybe we can use `def_equals_and_hash`?

File: src/socket/address.cr [hunk 4]
```diff
@@ -236,6 +236,8 @@ class Socket
       {% end %}
     end
 
+    def_hash family, port, address
```
Comment (asterite) [inline src/socket/address.cr:239]: We can use `def_equals_and_hash`

File: spec/std/socket/address_spec.cr [hunk 5]
```diff
@@ -127,6 +127,16 @@ describe Socket::IPAddress do
     Socket::IPAddress.valid_port?(-1).should be_false
     Socket::IPAddress.valid_port?(65_536).should be_false
   end
+  it "#==" do
```
Comment (asterite) [inline spec/std/socket/address_spec.cr:130]: Missing a newline before this

Comment (Blacksmoke16) [reply spec/std/socket/address_spec.cr:130]: Related https://github.com/crystal-lang/crystal/issues/9804.

**Review by asterite — APPROVED** [review]
Looks great!

**Pull request merged**